adaptive/learner/average_mixin.py

```diff
@@ -0,0 +1,202 @@
+        """Return n points that are expected to maximally reduce the loss."""
+        points, loss_improvements = self._ask_points_without_adding(n)
+        loss_improvements = self._normalize_new_points_loss_improvements(
+            points, loss_improvements)
+
+        p, l = self.loss_per_existing_point()
+        l = self._normalize_existing_points_loss_improvements(p, l)
+        points += p
+        loss_improvements += l
+
+        loss_improvements, points = zip(*sorted(
+            zip(loss_improvements, points), reverse=True))
+
+        points = list(points)[:n]
+        loss_improvements = list(loss_improvements)[:n]
+
+        if tell_pending:
+            for p in points:
+                self.tell_pending(p)
+
+        return points, loss_improvements
+
+    def n_values(self, point):
+        pending_points = self.pending_points.get(point, [])
+        return len(self._data[point]) + len(pending_points)
+
+    def _mean_values_per_neighbor(self, neighbors):
+        """The average number of neighbors of a 'point'."""
+        return {p: sum(self.n_values(n) for n in ns) / len(ns)
+            for p, ns in neighbors.items()}
+
+    def _normalize_new_points_loss_improvements(self, points, loss_improvements):
+        """If we are suggesting a new point, then its 'loss_improvement' should
+        be divided by the average number of values of its neigbors."""
+        if len(self._data) < 4:
+            return loss_improvements
+
+        only_points = [p for p, s in points]
+        neighbors = self._get_neighbor_mapping_new_points(only_points)
+        mean_values_per_neighbor = self._mean_values_per_neighbor(neighbors)
+
+        return [loss / mean_values_per_neighbor[p]
+            for (p, seed), loss in zip(points, loss_improvements)]
+
+    def _normalize_existing_points_loss_improvements(self, points, loss_improvements):
+        """If the neighbors of 'point' have twice as much values
+        on average, then that 'point' should have an infinite loss."""
+        if len(self._data) < 4:
+            return loss_improvements
+
+        neighbors = self._get_neighbor_mapping_existing_points()
+        mean_values_per_neighbor = self._mean_values_per_neighbor(neighbors)
+
+        def needs_more_data(p):
+            return mean_values_per_neighbor[p] > 1.5 * self.n_values(p)
+
+        return [inf if needs_more_data(p) else loss
+            for (p, seed), loss in zip(points, loss_improvements)]
+
+
+def add_average_mixin(cls):
+    names = ('data', 'data_sem', 'mean_values_per_point',
+             'get_seed', 'loss_per_existing_point', '_add_to_pending',
+             '_remove_from_to_pending', '_add_to_data', 'ask', 'n_values',
+             '_normalize_new_points_loss_improvements',
+             '_normalize_existing_points_loss_improvements',
+             '_mean_values_per_neighbor')
+
+    for name in names:
+        setattr(cls, name, getattr(AverageMixin, name))
+
+    return cls
+
+
+class DataPoint(dict):
```

Comment:
Isn't this an `AverageLearner`?

Reply:
Not really, it's similar.

It is really only a container that keeps track of the sum and sum of the squared values, such that it doesn't need to do the sum each time.

Reply:
We can use it there too actually. I've added it in the `AverageLearner`.